src/Symfony/Component/DependencyInjection/Compiler/PriorityTaggedServiceTrait.php

```diff
@@ -0,0 +1,47 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * Trait that allows a generic method to find and sort service by priority option in the tag.
+ *
+ * @author Iltar van der Berg <kjarli@gmail.com>
+ */
+trait PriorityTaggedServiceTrait
+{
+    /**
+     * Finds all services with the given tag name and order them by their priority.
+     *
+     * @param string           $tagName
+     * @param ContainerBuilder $container
+     *
+     * @return Reference[]
+     */
+    private function findAndSortTaggedServices($tagName, ContainerBuilder $container)
+    {
+        $services = $container->findTaggedServiceIds($tagName);
+
+        $queue = new \SplPriorityQueue();
+
```

Comment:
This PR made me think about handling priorities for formatters in #18450, but even if I love @javiereguiluz proposal to use `\SplPriorityQueue` (btw thanks for the heads-up :), I did not use it because I want to check that the same class is not used for many instances.

So I'm wondering even if the context is a bit different here, what about checking classes of tagged services to prevent one class to be registered with many service ids? Is this an expected possibility? Shouldn't we throw an exception in such case and add a test for it?

We could hold a `$classes = array()` before the `foreach` loops to do something like:
```php
$services = $container->findTaggedServiceIds($tagName);
$queue = new \SplPriorityQueue();
$classes = array();

foreach ($services as $serviceId => $tags) {
    $serviceClass = $container->findDefinition($serviceId)->getClass();
    if ($serviceClass && isset($classes[$serviceClass]) {
        throw new InvalidArgumentExcpetion(sprintf('The service "%s" has the same "%s" class as service "%s"', $serviceId, $serviceClass, $classes[$serviceClass]);
    }
    $classes[$attributes['class']] = $serviceId;
    foreach ($tags as $attributes) {
        $priority = isset($attributes['priority']) ? $attributes['priority'] : 0;
        $queue->insert(new Reference($serviceId), $priority * -1);
    }

    return iterator_to_array($queue);
}
```

What do you think ?